devito/types/tensor.py

```diff
@@ -0,0 +1,603 @@
+        name = '%s_T' % self.name
+        to = getattr(self, 'time_order', 0)
+        return func(name=name, grid=self.grid, space_order=self.space_order,
+                    components=self._mat, time_order=to, transpose=not self.is_transposed)
+
+    @property
+    def div(self):
+        """
+        Divergence of the VectorFunction, creates the divergence Function.
+        """
+        return sum([getattr(self[i], 'd%s' % d.name)
+                    for i, d in enumerate(self.space_dimensions)])
+
+    @property
+    def laplace(self):
+        """
+        Laplacian of the VectorFunction, creates the Laplacian VectorFunction.
+        """
+        comps = []
+        to = getattr(self, 'time_order', 0)
+        func = vec_func(self, self)
+        comps = [sum([getattr(s, 'd%s2' % d.name) for d in self.space_dimensions])
+                 for s in self]
+        return func(name='lap_%s' % self.name, grid=self.grid,
+                    space_order=self.space_order, components=comps, time_order=to)
+
+    @property
+    def curl(self):
+        """
+        Gradient of the (3D) VectorFunction, creates the curl VectorFunction.
+        """
+        if len(self.space_dimensions) != 3:
+            error("Curl only defined in three dimensions")
+        # The curl of a VectorFunction is a VectorFunction
+        derivs = ['d%s' % d.name for d in self.space_dimensions]
+        comp1 = getattr(self[2], derivs[1]) - getattr(self[1], derivs[2])
+        comp2 = getattr(self[0], derivs[2]) - getattr(self[2], derivs[0])
+        comp3 = getattr(self[1], derivs[0]) - getattr(self[0], derivs[1])
+
+        vec_func = VectorTimeFunction if self.is_TimeDependent else VectorFunction
+        to = getattr(self, 'time_order', 0)
+        return vec_func(name='curl_%s' % self.name, grid=self.grid,
+                        space_order=self.space_order, time_order=to,
+                        components=[comp1, comp2, comp3])
+
+    @property
+    def grad(self):
+        """
+        Gradient of the VectorFunction, creates the gradient TensorFunction.
+        """
+        to = getattr(self, 'time_order', 0)
+        func = tens_func(self, self)
+        comps = [[getattr(f, 'd%s' % d.name) for d in self.space_dimensions]
+                 for f in self]
+        return func(name='grad_%s' % self.name, grid=self.grid, time_order=to,
+                    space_order=self.space_order, components=comps, symmetric=False)
+
+    def new_from_mat(self, mat):
+        func = vec_func(self, self)
+        name = "%s%s" % ("_", self.name)
+        to = getattr(self, 'time_order', 0)
+        return func(name=name, grid=self.grid, space_order=self.space_order,
+                    components=mat, time_order=to, symmetric=self.is_symmetric,
+                    diagonal=self.is_diagonal)
+
+
+class VectorTimeFunction(VectorFunction, TensorTimeFunction):
+    """
+    Time varying VectorFunction.
+    """
+    is_VectorValued = True
```

Comment:
group these attribtues and use blank spaces , because the current way is just chaos